profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/extensions/ProfitBricksImageExtension.java

```diff
@@ -0,0 +1,146 @@
+import org.jclouds.compute.domain.ImageBuilder;
+import org.jclouds.compute.domain.ImageTemplate;
+import org.jclouds.compute.domain.ImageTemplateBuilder;
+import org.jclouds.compute.domain.OperatingSystem;
+import org.jclouds.compute.extensions.ImageExtension;
+import org.jclouds.compute.reference.ComputeServiceConstants;
+import org.jclouds.domain.Location;
+import static org.jclouds.location.predicates.LocationPredicates.idEquals;
+import org.jclouds.logging.Logger;
+
+public class ProfitBricksImageExtension implements ImageExtension {
+
+   @Resource
+   @Named(ComputeServiceConstants.COMPUTE_LOGGER)
+   protected Logger logger = Logger.NULL;
+
+   private final ProfitBricksApi client;
+   private final ListeningExecutorService userExecutor;
+   private final Supplier<Set<? extends Location>> locations;
+   private final Predicate<String> snapshotAvailablePredicate;
+   private final Trackables trackables;
+
+   @Inject
+   public ProfitBricksImageExtension(ProfitBricksApi client,
+           @Named(Constants.PROPERTY_USER_THREADS) ListeningExecutorService userExecutor,
+           @Memoized Supplier<Set<? extends Location>> locations,
+           @Named(POLL_PREDICATE_SNAPSHOT) Predicate<String> snapshotAvailablePredicate,
+           Trackables trackables) {
+      this.client = checkNotNull(client, "client");
+      this.userExecutor = checkNotNull(userExecutor, "userExecutor");
+      this.locations = checkNotNull(locations, "locations");
+      this.snapshotAvailablePredicate = checkNotNull(snapshotAvailablePredicate, "snapshotAvailablePredicate");
+      this.trackables = checkNotNull(trackables);
+   }
+
+   @Override
+   public ImageTemplate buildImageTemplateFromNode(String name, String id) {
+      DataCenterAndId datacenterAndId = DataCenterAndId.fromSlashEncoded(id);
+      Server server = client.serverApi().getServer(datacenterAndId.getDataCenter(), datacenterAndId.getId());
+      if (server == null) {
+         throw new NoSuchElementException("Cannot find server with id: " + id);
+      }
+      CloneImageTemplate template = new ImageTemplateBuilder.CloneImageTemplateBuilder().nodeId(id).name(name).build();
+      return template;
+   }
+
+   @Override
+   public ListenableFuture<Image> createImage(ImageTemplate template) {
+      checkState(template instanceof CloneImageTemplate,
+              " cloudstack only currently supports creating images through cloning.");
+      CloneImageTemplate cloneTemplate = (CloneImageTemplate) template;
+      DataCenterAndId datacenterAndId = DataCenterAndId.fromSlashEncoded(cloneTemplate.getSourceNodeId());
+
+      Server server = client.serverApi().getServer(datacenterAndId.getDataCenter(), datacenterAndId.getId());
+      List<Volume> volumes = client.volumeApi().getList(server.dataCenterId());
+
+      Volume volume = Iterables.getOnlyElement(volumes);
+
+      Snapshot snapshot = client.volumeApi().createSnapshot(Volume.Request.createSnapshotBuilder()
```

Comment:
Let's wrap all the volume snapshot creation inside the future, so this method can return early. If it returns a future, then, let's do all the job to create the snapshot asynchronously.
The operations to get the server and volume are ok outside the future, since they can be seen as validation and it's better if we can fail synchronously if the preconditions are not met.